### DISCUSSION TITLE
Breaking change: removing query parameters

### DISCUSSION THREAD

**Discussion by savage-alex** [opticdev/optic#1270] [Built-in Rulesets]
If I decide to rename or remove a query parameter, I think it should be warned as a breaking change. 

Views?

**Comment by acunniffe** [comment]
Hey @savage-alex  -- so I don't think this is technically a breaking change because a consumer would not have their clients suddenly stop sending valid requests in production. Unless your server explicitly disallows anything extra to be sent to it, this should not be breaking

However, I do think there are rules that make sense here. There's no reason your company could not define this as a violation of your versioning or deprecation policy by writing a rule about it. 

For instance I've seen teams write Optic rules that require query parameters have to have been marked `deprecated` for 90 days before being removed.

**Reply by savage-alex** [reply]
Love this proposition. We use quite specific validation so if we got something the server wasnt expecting we would tell the consumer they made a bad request. So for us it would be breaking. Just dropping some ideas here. Is there a "Warning" level response much like spectral saying you SHOULD NOT rather than you MUST NOT...

**Comment by acunniffe** [comment]
You could do something like this

```
queryParameter.removed(
  'removed query parameters must have been deprecated',
  (query) => {
    if (!query.value.deprecated)
      throw new RuleError({
        message: `Query parameters should not be deleted unless they have been deprecated`,
      });
  }
);
```
